build_modules/lib/src/module_builder.dart

```diff
@@ -7,15 +7,95 @@ import 'dart:convert';
 
 import 'package:build/build.dart';
 
+import 'meta_module.dart';
+import 'meta_module_builder.dart';
 import 'modules.dart';
 
 /// The extension for serialized module assets.
 const moduleExtension = '.module';
 
+/// A map of primary [AssetId] to corresponding [Module].
+final _primaryToClean =
+    new Resource<Map<AssetId, Module>>(() => {}, dispose: (map) => map.clear());
+
+/// Set of all [MetaModule]s which have been processed.
+final _readMetas = new Resource<Set<AssetId>>(() => new Set<AssetId>(),
+    dispose: (s) => s.clear());
+
+/// Map of [AssetId] to corresponding primary [AssetId] within the same
+/// [Module].
+final _assetToPrimary = new Resource<Map<AssetId, AssetId>>(() => {},
+    dispose: (map) => map.clear());
+
+void _cacheAssetToPrimary(
+    MetaModule meta, Map<AssetId, AssetId> assetToPrimary) {
+  for (var module in meta.modules) {
+    for (var source in module.sources) {
+      assetToPrimary[source] = module.primarySource;
+    }
+  }
+}
+
+/// Removes dependencies from the provided [Module] so that only those with
+/// corresponding module files remain.
+Future<Module> _cleanModuleDeps(
+    BuildStep buildStep, Module module, Map<AssetId, AssetId> assetToPrimary,
+    {bool isCourseStrategy: false}) async {
+  var cleanedDeps = new Set<AssetId>();
+  for (var dep in module.directDependencies) {
+    if (!isCourseStrategy) {
+      // Since we are not using the ccourse strategy we can safely add
+      // all dependencies in the same package as they will have a
+      // corresponding module file.
+      if (dep.package == module.primarySource.package) {
+        cleanedDeps.add(dep);
+        continue;
+        // It is possible that this dep came from another fine grained
+        // module. Look for the corresponding module file.
+      } else if (await buildStep
+          .canRead(dep.changeExtension(moduleExtension))) {
+        cleanedDeps.add(dep);
+        continue;
+      }
+    }
+    var primaryDep = assetToPrimary[dep];
+    if (primaryDep == null) {
+      // We don't know the primary source for the dep so look for it
+      // using the meta module file.
+      var depMeta = new MetaModule.fromJson(json.decode(
```

Comment:
Related to my comment above, but I believe this will fail if there is a module cycle between two packages. We can discuss later :D.